src/main/java/org/broadinstitute/hellbender/tools/walkers/featuremapping/AddFlowSNVQuality.java

```diff
@@ -0,0 +1,595 @@
+
+                // override result for the last base with the original hmer error probability
+                if ( base == result.length ) {
+                    result[base - 1] = errorProbBands[ERROR_PROB_BAND_KEY][flow];
+                }
+            }
+        }
+
+        // adjust probability of called bases so that sum will be 1, also enforce min prob
+        final byte[] bases = fbRead.getBasesNoCopy();
+        for ( int ofs = 0 ; ofs < bases.length ; ofs++ ) {
+
+            // go through alt bases and accumulate p, find out index of called bases (in flow order)
+            final byte calledBase = bases[ofs];
+            double altP = 0;
+            int calledIndex = -1;
+            for (int i = 0; i < flowOrderLength; i++) {
+                if ( calledBase != flowOrder[i] ) {
+                    snvResult[i][ofs] = Math.max(minLikelihoodProbRate, snvResult[i][ofs]);
+                    altP += snvResult[i][ofs];
+                } else {
+                    calledIndex = i;
+                }
+            }
+            if ( calledBase < 0 ) {
+                throw new GATKException(String.format("failed to locate called base %c in flow order %s", (char)calledBase, flowOrder));
+            }
+
+            // install probability in called base slot
+            snvResult[calledIndex][ofs] = Math.max(0, 1 - altP);
+
+            // at this point, bq becomes trivial (?)
+            result[ofs] = 1 - snvResult[calledIndex][ofs];
+        }
+
+        return result;
+    }
+
+    // extract error probability bands. middle (1) band is the key prob.
+    // lower (0) and high (2) are corresponding to -1 and +1 in hmer lengths
+    private static double[][] extractErrorProbBands(final FlowBasedRead flowRead, final double minValue) {
+
+        // access key
+        final int[] key = flowRead.getKey();
+
+        // allocate result
+        double[][] result = new double[ERROR_PROB_BANDS][];
+        for ( int i = 0 ; i < result.length ; i++ ) {
+            result[i] = new double[key.length];
+        }
+
+        for ( int i = 0 ; i < key.length ; i++ ) {
+
+            // extract key probability
+            result[ERROR_PROB_BAND_KEY][i] = Math.max(flowRead.getProb(i, key[i]), minValue);
+
+            // -1
+            if ( key[i] > 0 ) {
+                result[ERROR_PROB_BAND_1LESS][i] = Math.max(flowRead.getProb(i, key[i] - 1), minValue);
+            } else {
+                result[ERROR_PROB_BAND_1LESS][i] = minValue;
+            }
+
+            // +1
+            if ( key[i] < flowRead.getMaxHmer() ) {
+                result[ERROR_PROB_BAND_1MORE][i] = Math.max(flowRead.getProb(i, key[i] + 1), minValue);
+            } else {
+                result[ERROR_PROB_BAND_1MORE][i] = minValue;
+            }
+        }
+
+        return result;
+    }
+
+    @VisibleForTesting
+    protected double[] generateHmerBaseErrorProbabilities(final int[] key, final double[][] errorProbBands, final int flow,
```

Comment:
The haplotype caller works in flow space - without converting to base qualities (as it is faster this way). Right @ilyasoifer

Reply:
These function estimate the probability of each possible mismatch. They use the same idea as we in the haplotypecaller to score match of the read to a haplotype, but the function itself is different since there is no input haplotype.

@dror27 - since this really is the heart of the calculation, maybe it is possible to add more documentation to these functions?

Reply:
Done

Reply:
Just checking, this comment was added to `AddFlowBaseQuality` rather than here, right?